src/edu/upenn/cit594/processor/Processor.java

```diff
@@ -0,0 +1,94 @@
+    Reader<Map<String,Area>> areaReader;
+    Reader<List<ParkingViolation>> parkingViolationReader;
+    Map<Reader<?>, DataStore<?>> readerToDataStoreMap;
+    
+    public Processor(Reader<List<Property>> propertyReader, Reader<Map<String,Area>> areaReader, Reader<List<ParkingViolation>> parkingViolationReader) {
+        this.propertyReader = propertyReader;
+        this.areaReader = areaReader;
+        this.parkingViolationReader = parkingViolationReader;
+        readerToDataStoreMap = new HashMap<>();
+    }
+    
+    public int sumPopulations() {
+        int populationSum = 0;
+        Map<String, Area> areaMap = getData(areaReader);
+        
+        for (Map.Entry<String, Area> entry : areaMap.entrySet()) {
+            populationSum += entry.getValue().population;
+        }
+        System.out.println("sum: " + populationSum);
+        return populationSum;
+    }
+    
+    
+    public void calculateTotalFinesPerCapita() {
+        Map<String,Double> zipcodeToFineMap = new TreeMap<>();
+        Map<String, Area> areaMap = getData(areaReader);
+        List<ParkingViolation> violations = getData(parkingViolationReader);
+        for (ParkingViolation violation : violations) {
+            if (areaMap.containsKey((violation.zipcode))) {
+                if (zipcodeToFineMap.containsKey(violation.zipcode)) {
+                    double sumFine = zipcodeToFineMap.get(violation.zipcode) + violation.fine;
+                    zipcodeToFineMap.put(violation.zipcode, sumFine);
+                } else {
+                    zipcodeToFineMap.put(violation.zipcode, violation.fine);
+                }
+            }
+        }
+    
+       
+        for (Map.Entry<String, Double> entry : zipcodeToFineMap.entrySet()) {
+            double finesPerCapita = entry.getValue() / areaMap.get(entry.getKey()).population;
+            zipcodeToFineMap.put(entry.getKey(), finesPerCapita);
+        }
+        
+        //TODO: move formatting to UI class - new DecimalFormat("#.####").format(finesPerCapita)
+        
+        System.out.println(zipcodeToFineMap);
+        
+    }
+    
+    public double calculateAverageMarketValue(String zipcode) {
+        int countProperties = 0;
+        double sumMarketValue = 0.0;
+        List<Property> properties = getData(propertyReader);
+        for (Property property : properties) {
+            if (property.zipcode.equals(zipcode)) {
+                countProperties++;
+                sumMarketValue += property.marketValue;
+            }
+        }
+        System.out.println(sumMarketValue / countProperties);
+        
+        return sumMarketValue / countProperties;
+    }
+    
+    
+    /**
+     * Gets data from readerToDataStoreMap based on provided reader.
+     * Puts data from reader in readerToDataStoreMap if not already there.
+     * @param reader the reader to get data from
+     * @param <T>
+     * @return
+     */
+    private <T> T getData(Reader<T> reader) {
```

Comment:
Nice method. I would name this something else so it is not the same as `getData` on the `DataStore` class. I would call it `getReaderData` or `getExistingData`

Personally would rather write things out a little bit longer so that we don't need to suppress a warning (TAs might not look at it fondly)

```suggestion
    private <T> T getData(Reader<T> reader) {
         if (readerToDataStoreMap.containsKey(reader)) {
              return readerToDataStoreMap.get(reader).getData();
         }
         
         return reader.read();
    }
```

Reply:
Open to any names - I'm terrible at naming things. getReaderData sounds good.

The two issues with the method as written is this won't store the DataStore in the readerToDataStoreMap after having read. So there won't be an optimization in regards to memoization. It will also to still require us to explicitly cast to T! See next paragraph.

The cast seems dangerous, but I don't think there's a way around it because we're dealing with Lists and Maps of an unknown type in our readerToDataStore variable (which is declare as Map<Reader<?>, DataStore<?>>). Because DataStore itself accepts generic types (which it has to because we're having it store Lists and Maps), it could be a DataStore of any type. This DataStore is being stored as the value to the Reader, but the readerToDataStoreMap has no way of verifying that the Reader key is the same type as the DataStore value that is being passed to it. We know that it is because we are only putting in DataStore values that are coming from the Reader.read() method (which requires it to return the same type). So if we want to have memoization in the processor, a cast is needed at some point, either here in this method (which is localized to the temporary variable dataStore), or - more dangerously - in each of the calculate methods whenever we are attempting to pull from the readerToDataStore.

Here's a few links that I read about the issue:

- https://stackoverflow.com/questions/262367/type-safety-unchecked-cast
- https://www.edureka.co/community/22171/how-can-i-address-the-unchecked-cast-warnings
- https://codereview.stackexchange.com/questions/3130/how-can-i-avoid-unchecked-cast-warning-in-my-generic-recursive-iterator
- https://intellipaat.com/community/144/how-do-i-address-unchecked-cast-warnings

If you like, this version is a little clearer:
`private <T> T getData(Reader<T> reader) {
        @SuppressWarnings("unchecked")  // safe cast because the data in dataStore from the readers' read method can only be the same type as reader
        T data = (T) readerToDataStoreMap.computeIfAbsent(reader, currentReader -> reader.read()).getData();
        return data;
    }`